### PR TITLE
samples: net: big_http_download: Update to be usable in automated CI testing

### PR DESCRIPTION
**Pull request by pfalcon**
This application was initially intended to be a manual, interactive
    sample, run by a user. However, it's also a useful real-world
    integration test, to be run in automated CI systems. So, provide
    a config suitable for such a usage: use local host machine as a
    server (to not depend on availability and characteristics of an
    Internet connection), and have a concrete completion criteria
    (by limiting number of iterations; given that performance of TCP
    is not to be not too high, limit it to just 1 for now).

### CODE REVIEW

**Comment by pfalcon** [comment]
~~~
-:50: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#50: FILE: samples/net/sockets/big_http_download/src/big_http_download.c:39:
+    CONFIG_SAMPLE_BIG_HTTP_DL_URL;$
~~~

Hopefully, it makes sense here, as it follows indentation of string literals in nearby #ifdef sections.

**Comment by pfalcon** [comment]
Regarding the "samples: big_http_download: Allow to configure URL via Kconfig" commit: my original idea was to allow both URL **and** hash to be Kconfig'urable, i.e. allow to use arbitrary test URL (and content on it), but turned out, kconfiglib.py has some issues with handling strings which contain C escapes: https://github.com/zephyrproject-rtos/zephyr/issues/23917 . So, waiting for @ulfalizer to look into it, and so far allow to just mirror preconfigured test files on other severs (like local ones).

File: samples/net/sockets/big_http_download/Kconfig [hunk 1]
```diff
@@ -11,4 +11,10 @@ config SAMPLE_BIG_HTTP_DL_URL
 	help
 	  URL to download.
 
+config SAMPLE_BIG_HTTP_DL_NUM_ITER
+	int "Limit number of iterations"
+	default -1
```
Comment (jukkar) [inline samples/net/sockets/big_http_download/Kconfig:16]: 0 would feel more intuitive as a unlimited value, as value of 0 would not make sense here. Same applies of course also to <0 values which are not meaningful here. That would require small changes to while loop, perhaps easiest is to change it to `do {} while();`

File: samples/net/sockets/big_http_download/sample.yaml [hunk 2]
```diff
@@ -16,3 +16,9 @@ tests:
     extra_configs:
       - CONFIG_NET_SOCKETS_POSIX_NAMES=n
       - CONFIG_POSIX_API=y
+  # CI-friendly config, using local host machine as
+  # a server and finite number of iterations.
+  sample.net.sockets.big_http_download.ci:
+    extra_configs:
+      - CONFIG_SAMPLE_BIG_HTTP_DL_URL="http://192.0.2.2:8980/static/vmlinuz"
```
Comment (jukkar) [inline samples/net/sockets/big_http_download/sample.yaml:23]: Better would be to put the values to an overlay and just use that if you need this kind of special configuration. If the values are here, then we run this for every sanitycheck run which is probably not needed.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: samples/net/sockets/big_http_download/Kconfig (hunk 1 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/Kconfig:16]: > 0 would feel more intuitive as a unlimited value, as value of 0 would not make sense here.

Well, 0 means "zero iterations". Whether that makes sense to use is up to users. I also wouldn't agree that it's more intuitive, because 0 is again, zero times, and it follows convention of POSIX select()/poll() calls, which take 0, or positive number, or negative (commonly -1) as infinite delay.

Are you sure this request is not nit-picking? If you'd insist that value of 0 is used in Kconfig, then my fix on the source side would be to add:

~~~
if (num_iter == 0) {
    num_iter = -1;
}
~~~

File: samples/net/sockets/big_http_download/sample.yaml (hunk 2 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/sample.yaml:23]: > If the values are here, then we run this for every sanitycheck run which is probably not needed.

That's true, but the whole idea is to use sanitycheck as the CI driver, i.e. sanitycheck is used to build binaries in the build system, and then produced binaries are submitted for testing on the hardware. I can propose following solutions:

* Mark sample.net.sockets.big_http_download.ci as a "slow" config (if that's possible for just one config of a sample.yaml).
* Vice-versa, mark the original sample.net.sockets.big_http_download as "slow" (which would be true, considering it ~ infinite runtime).
* Remove the original sample.net.sockets.big_http_download config, we won't lose anything, as new "sample.net.sockets.big_http_download.ci" is about the same re: code coverage (even slightly better, as runs code after the loop).
* Just replace the original sample.net.sockets.big_http_download to use those options, which kinda makes sense as sanitycheck output is intended for automated testing, and the new configuration.

Hopefully, at least one of these options would work.

File: samples/net/sockets/big_http_download/Kconfig (hunk 1 above)
Comment (jukkar) [reply samples/net/sockets/big_http_download/Kconfig:16]: 0 zero value here would not make sense as you would not do any iterations then, so it is a special value. 
If using value 0 as limitless, then just do
```
do {

} while (--num_iterations);
```
below.

Comment (pfalcon) [reply samples/net/sockets/big_http_download/Kconfig:16]: > } while (--num_iterations);

Ok, updated as suggested.

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 3]
```diff
@@ -35,11 +35,15 @@
 
 /* This URL is parsed in-place, so buffer must be non-const. */
 static char download_url[] =
+#if defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL)
```
Comment (rlubos) [inline samples/net/sockets/big_http_download/src/big_http_download.c:38]: Won't this condition always be true? The config is added to the sample unconditionally.

Anyway, I think it would be neater to simply add the following line to the `overlay-tls.conf`:
```
CONFIG_SAMPLE_BIG_HTTP_DL_URL="https://www.7-zip.org/a/7z1805.exe"
```
and give up on these nested ifdefs.

Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: If we build with Zephyr build system. But that's not the only target and build system supported, this (and other samples I wrote) also work with standard POSIX make.

> Anyway, I think it would be neater to simply add the following line to the overlay-tls.conf:

Yes, the original idea was to make fully configurable via Kconfig. But that also requires configuring hash value, and as https://github.com/zephyrproject-rtos/zephyr/pull/24124#issuecomment-609791549 comment explains, that's blocked on #23917. Hopefully you agree, that moving URLs to Kconfig and leave hashes in .c would only make matters more confusing. (There's also a question of POSIX compat mentioned above, but as TLS support in this sample isn't POSIX-compliant, that can be discounted.)

**Comment by pfalcon** [comment]
@jukkar, can you please re-review?

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 3 above)
Comment (rlubos) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: Ok, let me rephrase, perhaps I wasn't clear enough.

My point is, that the current approach breaks the sample with TLS enabled. In Zephyr environment, the string provided by CONFIG_SAMPLE_BIG_HTTP_DL_URL will be used **unconditionally**, so the following will never be evaluated (in Zephyr):
```
#if !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
    "http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz";
#else
    "https://www.7-zip.org/a/7z1805.exe";
#endif /* !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
```

So we need to move the other address to Kconfig anyway (either by overlay or by a proper default in the Kconfig file).

I understand we need `else` for non-Zephyr environments, but it should be the sample default, in this case:
```
"http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz"
```

> Hopefully you agree, that moving URLs to Kconfig and leave hashes in .c would only make matters more confusing.

Yes I agree that might be confusing, but isn't this commit exaclty doing that? Anyway, I'd prefer some  (hopefully temporary) confusion over a broken sample.

Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: > CONFIG_SAMPLE_BIG_HTTP_DL_URL will be used unconditionally

Got it, thanks. My preference would be to move to the overlay config indeed.

**Comment by pfalcon** [comment]
@rlubos : Addressed regression with overlay-tls.conf pointed by you. Tested that both with and without overlay-tls.conf, the right URL is chosen. I however didn't make any further changes to the "defaults" in .c file, due to confusion it may lead to, explained in https://github.com/zephyrproject-rtos/zephyr/pull/24124#discussion_r404662143 . I hope https://github.com/zephyrproject-rtos/zephyr/issues/23917 will be resolved soon enough, and then we'll be able to cleanup this sample properly.

**Review by rlubos — APPROVED** [review]
Thank you.

File: samples/net/sockets/big_http_download/sample.yaml (hunk 2 above)
Comment (jukkar) [reply samples/net/sockets/big_http_download/sample.yaml:23]: If you place these settings to the overlay, and use the overlay here like this:
```
extra_args: OVERLAY_CONFIG=overlay-big-http-dl.conf
```
Then it is possible for the sanitychecker to use it, and if needed user can manually compile the sample with this configuration.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 3 above)
Comment (jukkar) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: The `defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL)` will match even if the value is empty string. So you need to add runtime check to make sure that the string is not empty.

File: samples/net/sockets/big_http_download/sample.yaml (hunk 2 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/sample.yaml:23]: > If you place these settings to the overlay, and use the overlay here like this:
extra_args: OVERLAY_CONFIG=overlay-big-http-dl.conf

Sounds good, made the change. I named the overlay `overlay-ci.conf` though (and moved comment from sample.yaml to the .conf).

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 3 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: > The defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL) will match even if the value is empty string. So you need to add runtime check to make sure that the string is not empty.

There's already such a check. The behavior with empty string is:

~~~
[00:00:02.380,000] <inf> net_config: Initializing network
[00:00:02.380,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:02.380,000] <inf> net_config: Running dhcpv4 client...
Error: Only http: URLs are supported
exit
~~~

Comment (jukkar) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: Ok, great. Also how these #ifdefs are now set means that the settings in #else branch
```
#if !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
    "http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz";
#else
    "https://www.7-zip.org/a/7z1805.exe";
#endif /* !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
```
 are never set, because the defined() only works properly for boolean types. So we could just remove these then.

Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: > are never set, because the defined() only works properly for boolean types.

I mentioned that already in a comment to @rlubos before, those settings are used when building without Zephyr Kconfig system, i.e. for native POSIX systems with `make -f Makefile.posix`.

Comment (jukkar) [reply samples/net/sockets/big_http_download/src/big_http_download.c:38]: Ok, sorry I missed it.

**Review by jukkar — APPROVED** [review]
